File: src/render-fragment.clj
(import* ["./base.js" "./prelude.clj" "./sanctuary.js"]

(let [replace (lambda [this that text] (.replace this that text))
      escape (pipe [(replace (regex "g" "&") "&amp;")
                    (replace (regex "g" "<") "&lt;")
                    (replace (regex "g" ">") "&gt;")])]

   (function render-fragment [indent level inline nodes]
      (++ (map (lambda [node]
                  (if (=== :text (:type node))
                      (escape (:value node))
                      (if (=== :excerpt (:type node))
                          (render-fragment indent level inline (:children node))
                          (let [indentation (.repeat level indent)
                                tag-name (symbol->string (:tag-name node))
                                attrs (++ (chain (lambda [sym]
                                                    [" "
                                                     (symbol->string sym)
                                                     "=\""
                                                     (escape (unwords (map trim (lines (sym (:attrs node))))))
                                                     "\""])
                                                 (.getOwnPropertySymbols (:attrs node) Object)))]
                             (++ (if (:self-closing node)
                                     [indentation "<" tag-name attrs " />" (if inline "" "\n")]
                                     (if (=== :inline (:format node))
                                         [indentation "<" tag-name attrs ">"
                                          (render-fragment indent 0 true (:children node))
                                          "</" tag-name ">" (if inline "" "\n")]
                                         [indentation "<" tag-name attrs ">\n"
                                          (render-fragment indent (add 1 level) false (:children node))
                                          indentation "</" tag-name ">\n"])))))))
               nodes)))))

File: src/render-fragment.js
'use strict';

const sanctuary = require ('sanctuary');


const {
  lines,
  map,
  trim,
  unwords,
} = sanctuary.unchecked;

const escape = s => (
  s
  .replace (/&/g, '&amp;')
  .replace (/</g, '&lt;')
  .replace (/>/g, '&gt;')
);

const symbolToString = sym => (
  String (sym)
  .slice ('Symbol('.length, -')'.length)
);

const $attrs = Symbol.for ('attrs');
const $children = Symbol.for ('children');
const $excerpt = Symbol.for ('excerpt');
const $format = Symbol.for ('format');
const $inline = Symbol.for ('inline');
const $selfClosing = Symbol.for ('self-closing');
const $tagName = Symbol.for ('tag-name');
const $text = Symbol.for ('text');
const $type = Symbol.for ('type');
const $value = Symbol.for ('value');

const renderFragment = module.exports = indent => level => inline => nodes => (
  nodes.reduce (
    (html, node) => {
      if (node[$type] === $text) {
        return html + escape (node[$value]);
      } else if (node[$type] === $excerpt) {
        return html +
               renderFragment (indent) (level) (inline) (node[$children]);
      } else {
        const indentation = indent.repeat (level);
        const tagName = symbolToString (node[$tagName]);
        const attrs = (
          Object.getOwnPropertySymbols (node[$attrs])
          .reduce (
            (s, sym) => (
              s +
              ' ' +
              symbolToString (sym) +
              '=' +
              '"' +
              escape (unwords (map (trim) (lines (node[$attrs][sym])))) +
              '"'
            ),
            '',
          )
        );
        if (node[$selfClosing]) {
          return html +
                 indentation + '<' + tagName + attrs + ' />' + (inline ? '' : '\n');
        } else if (node[$format] === $inline) {
          return html +
                 indentation + '<' + tagName + attrs + '>' +
                 renderFragment (indent) (0) (true) (node[$children]) +
                 '</' + tagName + '>' + (inline ? '' : '\n');
        } else {
          return html +
                 indentation + '<' + tagName + attrs + '>\n' +
                 renderFragment (indent) (level + 1) (false) (node[$children]) +
                 indentation + '</' + tagName + '>\n';
        }
      }
    },
    '',
  )
);
